src/network/cni/namespace.rs

```diff
@@ -0,0 +1,89 @@
+//! Network namespace helpers and structures.
+
+use anyhow::{Context, Result};
+use futures::executor;
+use getset::Getters;
+use log::trace;
+use nix::{
+    sched::{setns, CloneFlags},
+    unistd::gettid,
+};
+use std::{
+    future::Future,
+    os::unix::io::AsRawFd,
+    path::{Path, PathBuf},
+    process,
+};
+use tokio::{fs::File, task};
+
+#[derive(Debug, Getters)]
+/// A basic network namespace abstraction.
+pub struct Namespace {
+    #[get]
+    file: File,
+}
+
+impl Namespace {
+    /// Create a new namespace.
+    pub async fn new<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        Ok(Self {
+            file: File::open(&path).await.context("open namespace file")?,
+        })
+    }
+
+    /// Run a future inside this network namespace
+    pub async fn run<F>(&self, fun: F) -> Result<()>
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        trace!("Using file as network namespace: {:?}", self.file());
+        let fd = self.file().as_raw_fd();
+
+        trace!(
+            "Current thread network namespace: {}",
+            Self::current_thread_namespace_path().display()
+        );
+
+        task::spawn_blocking(move || {
+            setns(fd, CloneFlags::CLONE_NEWNET).context("switch to network namespace")?;
+            trace!(
+                "Switched to network namespace: {}",
+                Self::current_thread_namespace_path().display()
+            );
+
+            executor::block_on(fun).context("run namespace future")
+        })
+        .await
+        .context("spawn namespace thread")?
+        .context("run in namespace thread")
+    }
+
+    /// Retrieve the current network namespace path of the thread.
+    /// /proc/self/ns/net returns the namespace of the main thread, not of whatever thread this
+    /// thread is running on. Make sure we use the thread's net namespace since the thread is
+    /// switching around.
+    pub fn current_thread_namespace_path() -> PathBuf {
+        PathBuf::from(format!("/proc/{}/task/{}/ns/net", process::id(), gettid(),))
```

Comment:
Did you mean to do `readlink` here and return the result? This path by itself is not super useful to figuring out what the actual network namespace was (it basically is telling you what the pid and tid are -- useful information, but not everything you'd want to know).

Reply:
Yes thanks, I'll change it to `readlink`. I only need the path for obtaining the host network namespace. :+1: